src/infer_request.cc

```diff
@@ -481,12 +482,20 @@ InferRequest::Exec(const bool is_decoupled)
     // Additional round trip required for asking the stub process
     // to fill in the GPU tensor buffers
     if (has_gpu_tensor) {
+      AllocatedSharedMemory<GPUBuffersShm> gpu_buffers_shm =
+          shm_pool->Load<GPUBuffersShm>(
+              request_batch_shm_ptr->gpu_buffers_handle);
       AllocatedSharedMemory<bi::managed_external_buffer::handle_t>
           gpu_buffers_handle =
               shm_pool->Load<bi::managed_external_buffer::handle_t>(
-                  request_batch_shm_ptr->gpu_buffers_handle);
+                  gpu_buffers_shm.data_->buffers);
       try {
 #ifdef TRITON_ENABLE_GPU
+        if (!gpu_buffers_shm.data_->success) {
+          std::unique_ptr<PbString> error = PbString::LoadFromSharedMemory(
+              shm_pool, gpu_buffers_shm.data_->error);
+          throw PythonBackendException(error->String());
+        }
```

Comment:
Can the scope of the `#ifdef TRITON_ENABLE_GPU` be moved after this `if()` block to make the scope as small as necessary? Looks like everything in this block might be defined/used regardless of the ifdef - including `gpu_buffers_shm`, correct me if I'm wrong.

Though at a glance, I do find it a bit strange that some of the stuff like `has_gpu_tensor` and `gpu_buffers_shm` happen outside of the ifdef today.

Reply:
Also, is the `else` case of the ifdef just a no-op? Looks like the whole try block is inside the ifdef, so I'm a bit confused by the logic when `TRITON_ENABLE_GPU=OFF`.

Reply:
> Can the scope of the #ifdef TRITON_ENABLE_GPU be moved after this if() block to make the scope as small as necessary? Looks like everything in this block might be defined/used regardless of the ifdef - including gpu_buffers_shm, correct me if I'm wrong.

That's correct. Fixed.

>  I do find it a bit strange that some of the stuff like has_gpu_tensor and gpu_buffers_shm happen outside of the ifdef today.

They can be moved to ifdef but they are harmless since they don't use CUDA APIs. `has_gpu_tensor` will be simply false.

> Also, is the else case of the ifdef just a no-op?

Yes, since this special handling is only required for GPU tensors, if `#TRITON_ENABLE_GPU` is off it means that we cannot handle GPU tensors and there is no special handling required.